src/biogeochem/CropType.F90

```diff
@@ -41,6 +39,10 @@ module CropType
      character(len=20) :: baset_mapping
      real(r8) :: baset_latvary_intercept
      real(r8) :: baset_latvary_slope
+     real(r8), pointer :: sdates_thisyr           (:,:) ! all actual sowing dates for this patch this year
+     real(r8), pointer :: hdates_thisyr           (:,:) ! all actual harvest dates for this patch this year
+     integer , pointer :: sowing_count            (:)   ! number of sowing events this year for this patch
+     integer , pointer :: harvest_count           (:)   ! number of sowing events this year for this patch
```

Comment:
I gave this a shot and ran into a problem. Here's what I added for reading `sdates_thisyr` (similar for `hdates_thisyear`):
```fortran
call restartvar(ncid=ncid, flag=flag, varname='sdates_thisyr', xtype=ncd_double,  & 
     dim1name='pft', dim2name='mxgrowseas', switchdim=.true., &
     long_name='crop sowing dates for this patch this year', units='day of year', &
     scale_by_thickness=.false., &
     interpinic_flag='interp', readvar=readvar, data=this%sdates_thisyr)
```
But then I get the following error:
```
Abort with message NetCDF: Invalid dimension ID or name in file /glade/u/home/samrabin/ctsm/libraries/parallelio/src/c
lib/pio_nc.c at line 812
```
I'm assuming this is because `mxgrowseas` isn't in the restart file, right? Is there any way around this?

Reply:
Wrapping it like so seems to do the trick:
```fortran
call check_var_or_dim(ncid, 'mxgrowseas', is_dim=.true., exists=found)
if (found) then
```

Reply:
I don't think you want that check_var_or_dim call. Instead, I think you need to add an ncd_defdim call in restFileMod.F90, similar to this:

https://github.com/ESCOMP/CTSM/blob/8d9f988f94720cee94c48a21b1f8bc865c9d2409/src/main/restFileMod.F90#L546

Reply:
Oh wait, I get it: maybe you already added that defdim call but the problem came in reading an earlier restart file? If so, then something like your proposal makes sense – to do a check before making the restartvar call.

Reply:
But I think you should only do the check if `flag == 'read'`, so something like this (a bit ugly, I know):

```Fortran
  ! BACKWARDS_COMPATIBILITY(wjs, 2022-02-02) Even though restartvar can safely be called
  ! for a non-existent variable, it gives an error for a non-existent dimension. (We
  ! recently added the mxgrowseas dimension to the restart file.) So first check if the
  ! restart file has the necessary dimension before attempting to do the read.
  if (flag == 'read') then
     call check_dim(ncid, 'mxgrowseas', dimexist)
     if (dimexist) then
        docall = .true.
     else
        docall = .false.
     end if
  else
     docall = .true.
  end if

  if (docall) then
     call restartvar(..., varname='sdates_thisyr', ...)
  end if
```

Reply:
Alright, I've had a go at this and it seems to work; see commit ac01b14. I included a new function that does the `docall` check you suggested, since I'm doing that once for `mxgrowseas` and again for `mxharvests`.

Reply:
Awesome, this looks great. Thanks for introducing a shared subroutine to do this check.

I have a few minor comments on your latest commit, which I'm marking as optional because if you don't have time to do them I won't hold up the PR for them:

- [ ] [OPTIONAL] If you declare ncid as `intent(inout)` in your new subroutine then you should be able to get rid of the local `ncid_local` copy. (I'm not sure if it's reliably safe to make a copy of the `ncid` variable like you do here, so this would probably be good to change.)
- [ ] [OPTIONAL] You could use `check_dim` rather than `check_var_or_dim`; the latter is mainly intended for use by callers that could have either a dimension or variable name (such as the general-purpose `find_var_on_file` subroutine).
- [ ] [OPTIONAL] If you can come up with more descriptive variable names than `c` and `d`, that would be preferable.

Reply:
Re: `check_dim()` (which I only see in river routing files): Wouldn't it fail when the dimension doesn't exist? I'm not sure whether the call of `pio_inq_dimid()` itself would cause a failure, but if not it'd result in `dimlen /= value`, right?

Reply:
You're right that check_dim isn't called directly from anywhere else in CTSM, but it is what you want here; in CTSM, check_var_or_dim just delegates to either check_dim or check_var, and the implementation of check_dim differs from the one in RtmIO.F90:

https://github.com/ESCOMP/CTSM/blob/8d9f988f94720cee94c48a21b1f8bc865c9d2409/src/main/ncdio_pio.F90.in#L304-L348

While it's okay to be using check_var_or_dim here, it's just an unnecessary extra level of indirection.

Reply:
Good point; I missed that in my search because I was only looking in files ending with `.F90`.

Okay, those three tasks are done:
- d1ca586bf1f95345f8899ee5415a07542c5880af: ncid now inout in CallRestartvarDimOK().
- 7601ec88fe409484962693e136e0786379aa6f1d: More descriptive variable names.
-  948bcc9b22233d814068e78fcf9e3444d9168c2b: CallRestartvarDimOK now uses check_dim() instead of check_var_or_dim().

And I've redone the merges, hopefully correctly this time.

Reply:
Awesome, looks good; thanks for fixing those things; and the merge looks good now.